src/sam2CFN.ts

```diff
@@ -0,0 +1,238 @@
+    } else {
+        property = resourcesSpec.makeProperty(propertyTypes.pop()!);
+    }
+    property['Required'] = isRequired;
+    return property as awsData.Property;
+}
+
+function processDefinition(type: string, typeDef: any, awsResourcesSpec: any) {
+
+    // create and register the type
+    let typePropertyName = '';
+    let typeProperties: any;
+    let typeRequired: any;
+
+    let resource = resourcesSpec.makeType();
+
+    if (resourcesSpec.isPropertyTypeFormat(type)) {
+        awsResourcesSpec['PropertyTypes'][type] = resource;
+        typePropertyName = resourcesSpec.getPropertyTypePropertyName(type);
+        typeProperties = typeDef['properties'];
+        typeRequired = typeDef['required'];
+    } else {
+        awsResourcesSpec['ResourceTypes'][type] = resource;
+        typeProperties = typeDef['properties']['Properties']['properties'];
+        typeRequired = typeDef['properties']['Properties']['required'];
+    }
+
+    // process definition's properties
+    if (!!typeProperties) {
+        for (let propertyName of Object.keys(typeProperties)) {
+            let propertyDef = typeProperties[propertyName];
+            let propertyIsRequired = !!typeRequired ? !!~typeRequired.indexOf(propertyName) : false;
+
+            let propertyTypes = [];
+            // nested property type
+            if (propertyDef.hasOwnProperty('properties')) {
+                propertyTypes = [`${typePropertyName}.${propertyName}`];
+                processDefinition(`${type}.${propertyName}`, propertyName, awsResourcesSpec);
+
+            // direct or so it might seem ($ref ???)
+            } else {
+                propertyTypes = resolveTypes(propertyDef, `${typePropertyName}#${propertyName}`);
+            }
+            let resourceProperty = buildResourceProperty(propertyTypes, propertyIsRequired);
+            resource['Properties'][propertyName] = resourceProperty;
+            resource['AdditionalProperties'] = false;
+            let arnAttribute = buildResourceProperty(['String']);
+            (<any>resource)['Attributes'] = {Ref: arnAttribute};
+        }
+    }
+}
+
+function samResourcesSpecification(samSchema: any, customSpecification: any): awsData.AWSResourcesSpecification {
+    let awsResourcesSpec: awsData.AWSResourcesSpecification = {
+        PropertyTypes: {},
+        ResourceTypes: {}
+    };
+
+    // filter SAM schema to include just type definitions
+    let samSchemaTypeDefinitions = Object.keys(samSchema['definitions']);
+    samSchemaTypeDefinitions = samSchemaTypeDefinitions.filter((x) => !!~x.indexOf('::'));
+    // process SAM schema type definitions
+    for (let type of samSchemaTypeDefinitions) {
+      let samSchemaDef = samSchema['definitions'][type];
+      processDefinition(type, samSchemaDef, awsResourcesSpec);
+    }
+    // enhance with custom specification
+    awsResourcesSpec = mergeOptions(awsResourcesSpec, customSpecification);
+
+    return awsResourcesSpec;
+};
+
+
+program
```

Comment:
Do we have any tests for this?

Reply:
For testing purposes, I've had to split this into two components a CLI part, namely `sam2CFN.ts `, and a function library, namely `sam2CFNUtils.ts `. Please let me know if this is OK and/or you would like any additional tests added.

Reply:
Is the Utils tested for sanity?

Reply:
By *Utils* I believe you are referring to `sam2CFNUtils.ts` and if so, then please define *sanity* ... 
Basically both have tests, the tests regarding the library part should cover all the functionality used in the conversion process and the CLI has a unit test that validates that the parameters are actually being parsed and that the default values aren't used if the parameters have actually been specified on the command-line.
I was considering adding a test for the CLI that actually touched a file on the FS but I thought that perhaps it might be best to just mock the FS functions and check if they are being called with the correct arguments.
Anyway, just let me know how you would like to proceed with this.

Reply:
Ok, that's great thanks! Just wanted to check the functionality of the library part was fully covered.